src/client/Pipeline.cpp

```diff
@@ -241,18 +271,41 @@ void PipelineImpl::buildForAppendOrRecovery(bool recovery) {
              */
             if (stage != PIPELINE_SETUP_CREATE && stage != PIPELINE_CLOSE
                     && static_cast<int>(nodes.size()) < replication && canAddDatanode) {
-                if (!addDatanodeToPipeline(excludedNodes)) {
+                // Single data node case
+                bool added = false;
+                if (nodes.empty() && useRemoved) {
+                    if (storageID.length()) {
+                        LOG(INFO, "Pipeline: Adding back only datanode %s", removed.formatAddress().c_str());
+                        nodes.push_back(removed);
```

Comment:
oh. although I marked the line it didn't add it to the query...

At the end of `buildForAppendOrRecovery`
```
     /*
      * Update pipeline at the namenode, non-idempotent RPC call.
      */
     lb->setPoolId(lastBlock->getPoolId());
     lb->setBlockId(lastBlock->getBlockId());
     lb->setLocations(nodes);
     lb->setStorageIDs(storageIDs);
     lb->setNumBytes(lastBlock->getNumBytes());
     lb->setOffset(lastBlock->getOffset());
     filesystem->updatePipeline(*lastBlock, *lb, nodes, storageIDs);
     lastBlock = lb;
 }
```